### ISSUE TITLE
Model checker quickly gets very slow

### ISSUE THREAD

**Issue by rzach** [Carnap/Carnap#38]
The model checker takes about 60 seconds to verify this when the domain has 5 elements, and 120 seconds if you put in extensions that show the two sentences to be inequivalent (eg, A=0,1, P=2, O = [2,3],[2,4],[0,3],[1,4])
```
~~~{.CounterModeler .Simple counterexample-to="equivalence" system="thomasBolducAndZachFOL2019" points=2}
PS5.5 Ax(A(x) -> Ey(Az(P(z) -> O(z,y)) /\ O(x,y))), Ey(Az(P(z) -> O(z,y)) /\ Ax(A(x) -> O(x,y)))
~~~
```
See problem 6 on https://carnap.io/shared/rzach@ucalgary.ca/279%20F19%20Problem%20Set%205.md

Testing is inherently intractable, of course, but there is probably potential for optimization--eg, does the checker always test both A and B when checking A -> B, or only when A is true? Etc?

**Comment by gleachkr** [comment]
Huh. It shouldn't be that slow. 5 elements and 3 quantifiers is not that complicated. It looks like the main thing that matters is the size of the domain. Even with empty relations, there's still a pretty long pause.

The thing that makes this puzzling is that when you run the satisfaction check for that pair of sentences and for a similarly sized (or bigger) model, but do it in the REPL before compiling to JavaScript, it returns instantaneously. So, although the checking algorithm is pretty naive and could be improved, I'm not sure it's the only bottleneck.

I'll experiment a bit. If I turn anything up, I'll post it here.

**Comment by gleachkr** [comment]
OK. I've done some optimization, including the one you've suggested. This gets the inequivalence test for the problem case you found down to pretty much instantaneous. The big problem appears to have been the way that the universal closures of the sentences to check were being generated ~(basically, the universal closures are computed lazily - they aren't computed until the formula is used, even for subformulas. Working hypothesis is that this resulted in thousands of redundant and comparatively expensive operations when a formula was evaluated over a decently sized domain)~

To get the current performance, I had to drop the translation to universal closure. I'm going to see if I can find a more efficient way to keep that feature, and then I'll fire off an update.

**Comment by gleachkr** [comment]
OK, I've fixed the `universalClosure` function and deployed the result. The problem wasn't the laziness, as it turns out, but just a more straightforward bug leading to the insertion of vacuous quantifiers that dramatically blew up the the complexity of the check. Thanks again for pointing this issue out. Closed by 75165f122167dfd96e97f497051eb8492e65aa6d.
